README.md
# General
* Author: Shadi Akiki ( shadiakiki1986 at gmail dot com )
* Licensed under WTFPL
* [![Build Status](https://secure.travis-ci.org/shadiakiki1986/zboota-server.png)](http://travis-ci.org/shadiakiki1986/zboota-server)

# Description
* Server-side of zboota app
* It serves an API to the following websites
 * http://www.isf.gov.lb/ar/speedtickets
 * http://www.parkmeterlebanon.com/default/STATMENT_OF_ACCOUNT.aspx
* It manages user information and updates their data daily from the sites above

# Architecture
* php server
* mobile app developed with cordova
* json messages
* dynamodb table for user data
* AWS EC2 instance
 * serves API
 * updates data daily

# Note 1: Dynamodb reserved keywords
* http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ReservedWords.html

# Note 2: Failed builds on travis-ci
* Build 42 (https://travis-ci.org/shadiakiki1986/zboota-server/builds/63521468) on travis-ci seems to fail with an error about TLS
* I had thought to have solved this issue by using the known hosts field in the travis yml file
* so I tried to rebuild the last successful commit, a08093d, by branching it out to a branch I called ''lastsuccess''
* Commit a08093d had passed in build 41 (https://travis-ci.org/shadiakiki1986/zboota-server/builds/63161583)
* however, building it again, via the whitelist entry in the travis yml file, seems to fail now in build 43 (https://travis-ci.org/shadiakiki1986/zboota-server/builds/63524924)
* I''ll need to understand what is going on .. later


CHANGELOG.md
# Ver 0.0.3
2015-06-23
* home page status of parkmeterlebanon corrected
* added states to zboota notification email, i.e. new ticket, closed ticket, etc.
 * set sendNtofications to use simulation only so as to watch how this behaves over the next few days
 * will set simulation to false once I find this to have settled
* updated tests

2015-06-22
* added api hook to return the web availability ddb table + integrated it into home page

2015-06-20
* added to home page status the statuses of data feeder websites
* because parkmeterlebanon is offline, the sendNotifications.php script emailed all registered users
 * this is a false alarm of course
* as such, I added a breaker to make sure that this doesn''t happen again
* I also split out the sendNotificationsCore function into a class and added tests
* when Not available is set, also checking if website is originally available to set the force=true
* added dynamodb caching of website statuses so as not to have to check every retrieval

2015-06-16
* added notes on crontab requirement in INSTALL.md
* added statistic on cars last got in 24 hrs and 1 hr
* added uploading statistics on cars last got to dynamodb table for plotting as status (similar to travis-ci status)
* integrated zboota-statistics into www/api/statistics.php
* integrated zboota-statistics on nTotal and nNew from carsLastGetIn24 hrs into home page as a jqplot
* added statistics from bkp file and backfilled usage stats for display on zboota-server home page

2015-06-13
* moved tests of status to status indicators on the server home page (like travis-ci/status page)
* added some more CORS headers to get2.php

2015-06-11
* dropped all comment references to changing the root variable to development (was outdated since change to config.php in root folder)
* consolidated tests into phpunit test MiscTest.php
* also moved S3client tests to phpunit test

2015-06-02
* consolidated statistics in class and added weekly email
* sanity check on area and number in the get.php and get2.php api hooks
* refactored syncCore to use new function errorInLpns

2015-05-28
* added statistics on returning cars (i.e. anonymous users)

2015-05-22
* split out test user stuff into separate class
* added getpassword api hook for test user
* renamed deleteTestUser to testUserDelete

2015-05-21
* added deleteTestUser for the sake of automated testing of zboota-app

2015-05-19
* noticed that my travis-ci was very slow to re-build
 * read http://blog.travis-ci.com/2014-12-17-faster-builds-with-container-based-infrastructure/
 * learned that it may be because of my sudo: required
 * refactored include /etc/zboota...config.php to use a config file in the root folder
 * set sudo: false in travis yml file
* also, since I''m still getting the tls alert, I''m trying to add the ssh_known_hosts addon option to travis yml
 * http://docs.travis-ci.com/user/ssh-known-hosts/
* setting aws version to 2.7.* in composer.json didn''t help resolve the tls alert issue
 * setting to >=2.7.* since that''s the minimum version for which the passFail update works (check unit test)
 * actually setting to >=2.7.27 since the asterisk is not accepted as a version number in composer
* added composer self-update to travis yml file

2015-05-18
* added/improved some tests
* added travis.yml for CI with travis-ci.org
* incorporated composer for phpunit, aws (instead of AWS_PHAR)
* replaced several instances of putItem with updateItem
* after finally getting around the gnu tls error on travis-ci connecting to dynamodb (soln was to change from us-west-2 to us-east-1)
 * i ran the unit tests locally and corrected them by connecting to us-east-1 (after manually creating the tables)
 * i also added tests for newly created accounts
 * using aws >=8.4 so that the update of passfail works
 * travis only testing php 5.5 since that was irrelevant to the gnu tls issue

2015-05-17
* photos: moving from using local folder to S3 bucket + tested
* added backup photos script + tested
* added line output in log after each sync

2015-05-16
* added statistic scripts on live data (versus analyse.R on bkp files)

2015-05-15
* force update a car if it is available in the ddb but with old data
* split out uploadPhotoCore from uploadPhoto
 * also added uploadPhotoAsDataUrl
 * also simplified core
* added scipts listEmpty, etc
* removed trailing period from email sentence of forgot password and new email
* get2 is same as get but for angular get
 * I need to keep get because current apps use it
 * I should decomission it when no apps use it anymore
* loading photo as dataurl instaed of image because of javascript security exceptino in the latter case
* somehow I had accidentally removed the conversion of errors to json from login.php
 * I remember it was when I saw an error that the user was not supposed to see
 * so I judged that I should keep errors internal
 * but now I realize that some errors are supposed to bubble up to the user, such as wrong password, or unregistered email
* some CORS, angular http, jquery ajax, php crap

2015-05-13
* suppress html parsing warnings
* can upload photos and download them
* removed error returned as json and kept as internal server error
* removing empty fields before sending to dynamodb: made more general to any field, and not just the label, area, or number
* Changelog: removed double equal from dates
* new.php, update.php: adding isset(argc) to avoid the warning when used from html

2015-05-08
* not force refreshing data in case hp/y/t data missing
 * just remove dm before returning to user
* added test to see service availability
* api minor improvements of check variables at start to avoid notices
* furthered test_09_connectivity.sh

2015-05-07
* case of cached data modified with the mechanique info removed should not return mechanique results ... done
* case of timeout frmo isf, pml, dawlati servers => returmn 'Not available' to client instead of internal server error
* minor notice corrections
* added header message

lib/mapArea.php
<?php

# map
function mapAreaIsf() {
return array(	'B'=>'2:B',
		'G'=>'3:G',
		#'not coded'=>'1:not coded',
		'R'=>'4:R',
		'Z'=>'5:Z',
		'S'=>'6:S',
		'T'=>'7:T',
		'D'=>'8:D',
		'J'=>'9:J',
		'M'=>'10:M',
		'N'=>'11:N',
		'O'=>'12:O',
		#'40:Emirates', '14:Qatar', '39:Lathkia', '44:California',
		#'45:Foreign Plate', '19:Saudi Arabia', '20:Ruins',
		#'21:Without Restrictions', '22:Homos', '23:Idlib',
		#'24:Damascus', '25:Kuwait', '27:Sharjah', '28:UN',
		#'46:Deir al-zour', '41:Tartous', '34:Abou Dhabi', '49:Turkey',
		#'50:Raqaa', '51:Daraa', '55:Iraq', '26:Dubai', '30:Switzerland',
		#'31:Halab', '32:Hama', '33:Jordan', '35:BLP', '36:Syria',
		#'37:Bahrain', '38:Sweida', '43:Hasakeh', '47:Oman', '48:Tried'
);
}

function mapAreaPml() {
return array('B'=>"B/ب",
	'G'=>"G/ج",
	'D'=>"D/د",
	'R'=>"R/ر",
	'Z'=>"Z/ز",
	'S'=>"S/ص",
	'T'=>"T/ط",
	'J'=>"J/ع",
	'M'=>"M/م",
	'N'=>"N/ن",
	'O'=>"O/و",
	'C'=>"C/ق"
);
}
